### ISSUE TITLE
[Feature Request] Lazy Recursive Instantiation

### ISSUE THREAD

**Issue by Queuecumber** [facebookresearch/hydra#1413]
# 🚀 Feature Request

This request is for an API around "lazy"recursive instantiation where a config node with a `_target_` (and therefore any child as well) is marked to not be instantiated if the instantiation comes recursively from a parent. This is in contrast to the current `_recursive_: false` behavior which *does* instantiate the marked node but does not instantiate any children.

## Motivation

This came up when trying to build a generic config interface when used with Pytorch Lightning. `LightningModules` define a function `configure_optimizers` which is supposed to return the optimizer and lr_scheduler for the model. I wanted these to be pluggable, so I define them in configs, for example this one for Adam:

config/model/optimizer/adam.yaml:
```
_target_: torch.optim.Adam
```

then my models can declare what optimizer to use (or the caller can override it):

config/model/qgac_crab.yaml:
```
defaults:
  - optimizer: adam
  - scheduler: cosine_annealing

_target_: mm.models.QGACCrab
stats: 
  _target_: torchjpeg.dct.Stats
  root: ${modpath:}/../stats/cstats.pt
``` 

or `python -m train model.optimizer=sgd`

However, the optimizer contstructor needs the model parameters to be instantiated. The problem is that my top level `instantiate` call, which is several levels removed from the model code, doesn't know anything about the underlying model (since it's an abstraction) and at any rate it's trying to instantiate the entire tree at once so it can't access the model parameters. I could add `_recursive: false` on the *model* config, but then the other model parameters, like `stats` in this case, wouldn't get instantiated. It's OK for now but it may be a problem if there are other complex trees and parameters that need to be instantiated to move that burden into the model constructor. Ideally my model constructor should look like:

```
class QGACCrab(pl.LightningModule):
    def __init__(self, stats: Stats, optimizer: DictConfig, scheduler: DictConfig) -> None:
        self.stats = stats
        ...
        optimizer = instantiate(self.optimizer_config, params=self.parameters())
        scheduler = instantiate(self.scheduler_config, optimizer=optimizer)
```

note that the Stats class was instantiated but the optimizer and scheduler were passed as dictconfig and instantiated my the model.

## Pitch

**Describe the solution you'd like**
I'd like to start a discussion about the "lazy" instantiation API where, in my example, the optimizer and scheduler config can declare that they should not be instantiated as a result of a recursive instantiation from a parent. Something like:

config/model/optimizer/adam.yaml:
```
_target_: torch.optim.Adam
_lazy_: yes
```

Note the subtle difference in meaning here: adam is saying "don't instantiate me (or my children)" vs:

config/model/qgac_crab.yaml:
```
...
_target_: mm.models.QGACCrab
stats: 
  _recursive_: no
  _target_: torchjpeg.dct.Stats
  root: ${modpath:}/../stats/cstats.pt
```

Which is saying "instantiate me but dont instantiate my children"

**Describe alternatives you've considered**
The only obvious alternative is using the current `_recursive_: false` API and moving all instantiation for the model node parameters into the model init.

**Are you willing to open a pull request?** (See [CONTRIBUTING](../../CONTRIBUTING.md))
Yes

## Additional context
None

**Comment by omry** [comment]
Thanks, adding tentatively to 1.1.
There are some other work around instantiation and I will try to squeeze this in as well, but no promises.

**Comment by Queuecumber** [comment]
OK let's see what we can come up with, if we can agree to an interface I'm happy to turn around the code quickly

**Comment by omry** [comment]
The instantiation code is subject to some major changes, so I would rather not add additional features before it settles down.

**Comment by omry** [comment]
@Queuecumber , can you take a look at #1283?
Can partial instantiation help in your use case?

**Comment by Queuecumber** [comment]
Absolutely and I like that solution more, I think it's more versatile 

However I'm in favor of implementing the `_partial_` flag as described in the issue, I think it would make the overall configuration a lot cleaner to write and read and i don't think the explanation of partial application is that complicated for a reader to understand.

How about we revisit this after the 1.1 launch?

**Comment by omry** [comment]
Yes. I reopened the other one for 1.2.
